src/engine_serializer.rs

```diff
@@ -0,0 +1,54 @@
+use crate::engine::Engine;
+
+#[derive(Debug)]
+pub enum SerializeError {
+    DataFormatError(crate::data_format::SerializationError),
+}
+
+#[derive(Debug)]
+pub enum DeserializeError {
+    DataFormatError(crate::data_format::DeserializationError),
+}
+
+impl From<crate::data_format::SerializationError> for SerializeError {
+    fn from(value: crate::data_format::SerializationError) -> Self {
+        SerializeError::DataFormatError(value)
+    }
+}
+
+impl From<crate::data_format::DeserializationError> for DeserializeError {
+    fn from(value: crate::data_format::DeserializationError) -> Self {
+        DeserializeError::DataFormatError(value)
+    }
+}
+
+pub trait EngineSerializer {
+    fn serialize(&self) -> Result<Vec<u8>, SerializeError>;
+    fn deserialize(&mut self, serialized: &[u8]) -> Result<(), DeserializeError>;
+}
+
+impl EngineSerializer for Engine {
```

Comment:
could we leave these directly on `impl Engine`? I see no reason to move it into a separate trait; just makes the API slightly more confusing for users